### DISCUSSION TITLE
PV Power Sensor?

### DISCUSSION THREAD

**Discussion by NigelGreenslade** [tomasmcguinness/homeassistant-mixergy#13] [Q&A]
On the Mixergy app 'Graphs' tab there's a sub-tab for showing PV Power. Is this or can this be made available as a sensor on this integration?

**Comment by tomasmcguinness** [comment]
My integration already provides power and energy based on the element being on/off. These feed into the “Energy” component of Home Assistant. Does this help?

**Reply by NigelGreenslade** [reply]
Not sure ... It's my understanding that the Mixergy's 'Solar tank' has a twin element which allows for separate heating using the PV Power derived from the Solar Diverter. I think your sensors 'mixergy_electric_heat_energy' and 'mixergy_electric_heat_power' only provide energy and power data for a standard tank (single element). Although I might be wrong?

**Reply by tomasmcguinness** [reply]
My tank has a dual element. One powered by the Mixergy and one powered by a solar diverter. I use both elements, using an Eddi, but the Mixergy graph doesn’t show any of that data. 

What setup have you got?

**Reply by NigelGreenslade** [reply]
I've got a Mixergy solar diverter (bolted onto the cylinder, but separate from the controller), which is probably why the data shows up on my Mixergy app.  Here a screenshot.

![Screenshot of Mixergy Graph tab (on App)](https://github.com/tomasmcguinness/homeassistant-mixergy/assets/119896304/4d2705b6-eb0b-4dba-a433-89326a3f8082)

**Reply by tomasmcguinness** [reply]
Ah, okay. They might provide that data in the API, but I don’t know what it would be. Home Assistant works with power and energy in a specific way, so I wouldn’t be able to use the graph data directly. I will reach out to Mixergy to see if the API returns that data.

**Reply by NigelGreenslade** [reply]
Please let me know once you've assessed if it is doable (or not).
Thank you and much appreciated.

**Reply by NigelGreenslade** [reply]
Hi, just wondering if you had any luck hearing back from Mixergy?

**Reply by tdroza** [reply]
Just to add my support to this thread, I'd really like to be able to get the Solar PV data from my Mixergy into HA. I don't have the API documentation but I took a look at the data that comes back from the `/measurements` endpoint when I view the graph on the Mixergy website and I couldn't see any numbers that seemed to correspond to my clamp or PV power values.

**Reply by tomasmcguinness** [reply]
Sorry all - this discussions panel is nothing something I've checked recently. 

I did ask, but didn't get anything back from them. 

If I knew what the API returned, I can easily add some additional sensors into the mix.

I'll reach out again!

**Comment by NigelGreenslade** [comment]
I reached out to Mixergy back in July and again 6 weeks ago and they replied saying this query wasn't their current priority. I suppose the more people who ask about this the more likely they'll increase its priority ...

I'll send them another email ...

**Comment by NigelGreenslade** [comment]
Hi Tomas. I've just received an email from Mixergy. Please let me know if this is helpful? See below:



“The measurement JSON has the following fields relating to PV and the current clamp: 

    clampPower:    number;
    clampCurrent:  number;
    pvEnergy:      number;

I think these are correct.
Clamp power is the average power in watts measured over the minute with the current clamp. This clamp is usually attached to the property supply at the consumer unit. Positive is an import, negative is an export.
Clamp current is the average current in amps over the last minute.
PV Energy is the PV energy in joules that is consumed by the tank. We don't have visibility of what the PV system is doing as a whole.”

**Reply by tomasmcguinness** [reply]
Ha. I got the same email earlier :)

I guess the clamp stuff isn't of that much interest. You want pvEnergy? 

My issue is that I can't test it. I don't even know what it really contains. I'm assuming it's the instantaneous power as opposed to energy.

How comfortable are you with running a python script?

**Reply by NigelGreenslade** [reply]
Yes, I'm just after the pvEnergy.

I assume its instantaneous power in Watts ...

I've never run a python script from HA before, so no.

**Reply by tomasmcguinness** [reply]
Would you be willing?

I've done this with one or two other users when I'm trying to get an idea of what the API returns for tank features I don't have (like heat pump support)

**Reply by NigelGreenslade** [reply]
Yep, I'm willing

**Reply by tomasmcguinness** [reply]
If you download the two files in this folder

https://github.com/tomasmcguinness/homeassistant-mixergy/blob/main/testing

Then, assuming you have Python installed, you can run 

```
pip install -r requirements.txt 
```

This installs a library for making http requests

Once that's installed, execute the script, supplying your username and password.

```
python mixergy.py <username> <password>
```

If this runs correctly, you should see some output around the measurements from the tank, the last one being PV Energy.

If that works, we'll know what's in there. If it doesn't, we'll try again!

**Reply by NigelGreenslade** [reply]
Hi. Just run the script and it just comes up with "** Found a tank with serial number MX010xxx"

**Reply by NigelGreenslade** [reply]
obviously I replaced the tank serial number with xxx

**Reply by tomasmcguinness** [reply]
Right. The serial number is the 3rd argument that I forgot about :)

```
python mixergy.py <username> <password> <serial>
```

Try that :)

**Reply by NigelGreenslade** [reply]
Hi. it shows output but no PV energy. The last 4 lines are:

Bottom Temp: 25.7
Charge: 74.8
Heat Source: Indirect
Heat Source On: False

**Reply by tdroza-nhs** [reply]
Just to note that I'm following this thread with interest. I've also run the python script and can confirm the same as @NigelGreenslade above. I've also output the raw json and I'm not seeing `clampPower` or `pvEnergy` in the data.

**Reply by tdroza-nhs** [reply]
It's possible that `pvEnergy` is ommitted from the data if the value is zero (as mine currently is, given that it's dark now). When I view the "MEasurements" tab on the mixergy website, the PV Energy column either has a positive value, or is empty (not zero).

**Reply by tdroza-nhs** [reply]
Ah yes, if I inspect the response to the `/measurements` API that the Mixergy website makes when I view the Measurements tab, I get back a list of 1039 measurements. 143 of those contain `pvEnergy` with a non-zero value. 

E.g.
```
        {
            "recordedTime": 1699438560003,
            "receivedTime": 1699438594737,
            "topTemperature": 42.8,
            "bottomTemperature": 10.4,
            "charge": 22.3,
            "frequency": 49.8947,
            "energy": 9883,
            "state": "{\"system\":\"On\",\"heat_source\":\"HeatPump\",\"current\":{\"immersion\":\"Off\",\"heat_source\":\"HeatPump\"},\"relay\":{\"immersion\":\"Off\",\"heat_source\":\"HeatPump\"}}",
            "voltage": 239.48,
            "current": 1.08,
            "clampPower": -334.83,
            "pvEnergy": 9883
        },
```

**Reply by tomasmcguinness** [reply]
Brilliant. You're most likely right saying it's omitted when 0. Does the value of pvEnergy tally with a value in Watts?

**Reply by tomasmcguinness** [reply]
@NigelGreenslade Do you think you might get some solar diverting tomorrow? I know it's the wrong time of the year for such things, but at least we know the data is there now.

**Reply by NigelGreenslade** [reply]
Unlikely. Looking at the Solar forecast the next 7 day, Saturday is te best, so maybe then?

**Reply by NigelGreenslade** [reply]
Problem is that I'd need keep a watch for when it actually happens, otherwise we'd miss it.

**Reply by tdroza-nhs** [reply]
No it's not watts - I guess the distinction is that the field is "pvEnergy" rather than "power". This is how the graph reports the same values that I pasted above in JSON
![image](https://github.com/tomasmcguinness/homeassistant-mixergy/assets/83403414/c92b859b-093a-436d-8907-d1b4fae263d4)

Digging through the JS from the Mixergy site I can see that they convert energy to power using `pvEnergy / 6e4`.

i.e. using my data from above, 9833/60000=0.16

**Reply by tomasmcguinness** [reply]
huh, energy is normally kW etc. I wonder what the division does? Anyway - I can factor that into my integration.

**Reply by tomasmcguinness** [reply]
https://github.com/tomasmcguinness/homeassistant-mixergy/releases/tag/v0.5.0

It's a pre-release version. Let me know if it works :)

**Reply by tdroza-nhs** [reply]
Not working for me unfortunately. I get the following error in the logs. I tried removing and readding the integration but same issue (have also restarted HA after updating to this version):
```
Error while setting up mixergy platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 359, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/mixergy/sensor.py", line 46, in async_setup_entry
    new_entities.append(PVEnergySensor(coordinator, tank))
                        ^^^^^^^^^^^^^^
NameError: name 'PVEnergySensor' is not defined
```

**Reply by tomasmcguinness** [reply]
This is why it's pre-release :)

**Reply by tomasmcguinness** [reply]
0.5.1 has been published. I didn't update the name of the sensor.

**Reply by tdroza-nhs** [reply]
It now loads correctly, but the `electric_pv_power` sensor is showing zero (I'm currently diverting 0.96kW into my tank so it shouldn't be zero). The response to the `/measurements` call from the Mixergy website's graph tab correctly shows the PV value, but when I run your test Python file it isn't returned. Could it be that it's only returned in the full measurements list and not in the "latest" call?

**Reply by tomasmcguinness** [reply]
If you run the current measurement endpoint, do you see anything in the JSON?

**Reply by tdroza-nhs** [reply]
If I debug `latest_measurement_result` from your test script I do see `pvEnergy` as the last field:

```
{'recordedTime': 1699525380063, 'receivedTime': 1699525387790, 'topTemperature': 59.9, 'bottomTemperature': 44.5, 'charge': 91.7, 'frequency': 49.9995, 'energy': 31210, 'state': '{"system":"On","heat_source":"HeatPump","current":{"immersion":"Off","heat_source":"HeatPump"},"relay":{"immersion":"Off","heat_source":"HeatPump"}}', 'voltage': 237.39, 'current': 3.07, 'clampPower': -249.06, 'pvEnergy': 31210}
```

**Reply by tomasmcguinness** [reply]
I think I'm looking for the data in the wrong place.

**Reply by tdroza-nhs** [reply]
Was just about to say the same - I think you're looking in the `current` object but it's outside of that

**Reply by tomasmcguinness** [reply]
0.5.2 is published

**Reply by tdroza-nhs** [reply]
That's working 👍 ....but the unit is wrong, It should either be kW or multiplied by 1000 to be W. (I may have misled you yesterday when talking about the conversion from energy to power).
![image](https://github.com/tomasmcguinness/homeassistant-mixergy/assets/83403414/281e5ad5-01e6-4116-a92f-193aa0a4d464)

**Reply by tomasmcguinness** [reply]
I've switched the unit to KILO_WATT (I think) - 0.5.3

**Reply by tdroza-nhs** [reply]
Awesome, that's working for me now! Really appreciate you taking the time to do this 👏 
![image](https://github.com/tomasmcguinness/homeassistant-mixergy/assets/83403414/a02d2bd7-cd8e-4e24-9569-a24c111a5aa3)
(updated with a better screenshot after my solar inverter data had refreshed to match the mixergy data)

**Reply by tomasmcguinness** [reply]
Ha! Brilliant.

**Reply by NigelGreenslade** [reply]
Brilliant! I see the new sensor, just waiting for some sun  ...
Many thanks for sorting this. Much appreciated

**Reply by tomasmcguinness** [reply]
All donations are appreciated https://ko-fi.com/tomasmcguinness ;)
